packages/headless/src/features/query/query-slice.ts

```diff
@@ -17,4 +18,7 @@ export const queryReducer = createReducer(getQueryInitialState(), (builder) =>
     .addCase(change.fulfilled, (state, action) => {
       state.q = action.payload.query.q;
     })
+    .addCase(restoreState, (state, action) => {
```

Comment:
The logic for restoring facet values from a url is different from restoring via a snapshot.

In the case of a snapshot, we restore exactly what is in the snapshot. If there are missing facets, new facets that are added, we put in state the exact snapshot. The reducer logic is simply:
```
.addCase(change.fulfilled, (_, action) => action.payload.facetSet)
```

When restoring from url, the payload will either be `Record<string, FacetValue[]>`, where the string key is the field name, or a partial of the facet request as you said, `{field: string, currentValues: [...]}`. However, the reducer logic to restore this is completely different.

```
.addCase(restoreState, (state, action) => {
    // loop through the facet requests we have in state
    // if the field of a facet request matches one in the payload, ensure the values in the payload exist on the request and are selected, while respecting the total number of values/adjusting if needed
   // else if the facet request field does not match one in the payload, ensure all values are unselected.
   // set preventAutoSelect to true so that DNE does not return additional selected values in the response
```

When restoring from url, it is not possible to just return the payload like in the case of the snapshot, or to destructure the payload on an existing facet request.

Reply:
Alright, fair enough !